pinot-druid/src/main/java/org/apache/pinot/druid/data/readers/DruidSegmentRecordReader.java

```diff
@@ -0,0 +1,219 @@
+import org.apache.druid.segment.QueryableIndex;
+import org.apache.druid.segment.QueryableIndexStorageAdapter;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ColumnConfig;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.filter.Filters;
+import org.apache.druid.query.DruidProcessingConfig;
+import org.apache.pinot.common.data.FieldSpec;
+import org.apache.pinot.common.data.Schema;
+import org.apache.pinot.core.data.GenericRow;
+import org.apache.pinot.core.data.readers.RecordReader;
+import org.apache.pinot.core.indexsegment.generator.SegmentGeneratorConfig;
+import org.joda.time.chrono.ISOChronology;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The DruidSegmentRecordReader allows us to convert all of the rows in a Druid segment file
+ * into GenericRows, which are made into Pinot segments.
+ *
+ * Note that Druid uses LONG and not INT, so construct the Pinot schema accordingly.
+ */
+public class DruidSegmentRecordReader implements RecordReader {
+  // TODO: Figure out why logger stuff isn't showing up in console
+  private static final Logger LOGGER = LoggerFactory.getLogger(DruidSegmentRecordReader.class);
+
+  private Schema _pinotSchema;
+  private Cursor _cursor;
+  private ArrayList<String> _columnNames;
+  private List<BaseObjectColumnValueSelector> _selectors;
+  private QueryableIndex _index;
+  private QueryableIndexStorageAdapter _adapter;
+
+  private void validateColumns()
+      throws IllegalArgumentException {
+    for (int i = 0; i < _columnNames.size(); i++) {
+      final String columnName = _columnNames.get(i);
+      ColumnCapabilities capabilities = _adapter.getColumnCapabilities(columnName);
+      if (capabilities == null) {
+        LOGGER.warn("Column {} is not in record", columnName);
+      } else {
+        if (capabilities.getType() == ValueType.COMPLEX) {
+          throw new IllegalArgumentException("Column " + columnName
+              + ": DruidSegmentRecordReader does not support complex metric columns.");
+        }
+        FieldSpec fieldSpec = _pinotSchema.getFieldSpecFor(_columnNames.get(i));
+        if (fieldSpec.isSingleValueField() && capabilities.hasMultipleValues()) {
+          throw new IllegalArgumentException("Column " + columnName
+              + " in Pinot schema is single-valued, but column {} in " + "record is multi-valued.");
+        }
+        if (!fieldSpec.isSingleValueField() && !capabilities.hasMultipleValues()) {
+          throw new IllegalArgumentException("Column " + columnName
+              + " in Pinot schema is multi-valued, but column {} in" + "record is single-valued");
+        }
+        String typeFromSchema = _pinotSchema.getFieldSpecFor(columnName).getDataType().toString();
+        String typeFromRecord = capabilities.getType().toString();
+        if (typeFromRecord != typeFromSchema) {
```

Comment:
1. You need to use `typeFromRecord.equals(typeFromSchema)` to compare the string. `!=` will only check if they are the same reference. https://stackoverflow.com/questions/513832/how-do-i-compare-strings-in-java
2. Pinot will give you `DataType` and Druid will give you `ValueType`. Instead of comparing `String` value directly, I think that it's better to add a compare function something like the following:

```
private boolean compareType(DataType pinotColumnType, ValueType druidColumnType) 
```
You can handle the following type equality within the function.

```
 public enum DataType {
    INT, LONG, FLOAT, DOUBLE, BOOLEAN/* Stored as STRING */, STRING, BYTES;
  }
```

```
public enum ValueType
{
  FLOAT, DOUBLE, LONG, STRING, COMPLEX 
}
```